crud/src/main/java/com/redhat/lightblue/assoc/AnalyzeQuery.java

```diff
@@ -0,0 +1,330 @@
+        // copy the indexes from the normalizedFieldName
+        FieldTreeNode trc=fieldNode;
+        int n=normalizedFieldName.numSegments()-1;
+        ArrayList<String> list=new ArrayList<>(n);
+        while(!((trc instanceof ArrayElement&&
+                ((ArrayElement)trc).getParent() instanceof ResolvedReferenceField) ||
+                trc.getParent()==null )) {
+            String name=trc.getName();
+            if(Path.ANY.equals(name)) {
+            	if(n>=0) {
+            		String head=normalizedFieldName.head(n);
+            		if(!head.equals("$parent")) {
+            			list.add(head);
+            		} else {
+            			list.add(Path.ANY);
+            		}
+            	} else {
+            		list.add(Path.ANY);
+            	}
+            } else {
+                list.add(name);
+            }
+            n--;
+            trc=trc.getParent();
+        } 
+        MutablePath p=new MutablePath();
+        n=list.size();
+        for(int i=n-1;i>=0;i--)
+            p.push(list.get(i));
+        return p.immutableCopy();
+    }
+
+    /**
+     * Returns true if 'ancestor' is an ancestor of 'descendant'. Does not cross entity boundaries.
+     */
+    private boolean isAnAncestor(FieldTreeNode ancestor,FieldTreeNode descendant) {
+        FieldTreeNode trc=descendant;
+        do {
+            if(trc==ancestor) {
+                return true;
+            } else {
+                trc=trc.getParent();
+            }
+        } while(trc.getParent()!=null&&!(trc instanceof ArrayElement &&
+                                         ((ArrayElement)trc).getParent() instanceof ResolvedReferenceField ));
+        return false;
+    }
+
+
+    private static class NearestCommonNode {
+        final FieldTreeNode node; // The nearest node, or null of root
+        final int numSegments; // Number of segment we have to back in context
+
+        public NearestCommonNode(FieldTreeNode node,int n) {
+            this.node=node;
+            this.numSegments=n;
+        }
+    }
+    
+    /**
+     * If contextNode and fieldNode are both in the same entity,
+     * returns the nearest common node in the context's ancestors, or null if there is none. 
+     * Also returns the number of nodes passed to get there.
+     */
+    private NearestCommonNode findNearestCommonNode(FieldTreeNode contextNode,FieldTreeNode fieldNode) {
+        int n=0;
+        FieldTreeNode trc=contextNode;
+        do {
+            if(isAnAncestor(trc,fieldNode)) {
+                return new NearestCommonNode(trc,n);
+            } else {
+                trc=trc.getParent();
+                n++;
+            }
+        } while(trc.getParent()!=null&&!(trc instanceof ArrayElement &&
+                                         ((ArrayElement)trc).getParent() instanceof ResolvedReferenceField ));
+        return new NearestCommonNode(null,n);
+    }
+    
+    private QueryFieldInfo resolveField(Path clauseFieldName,
```

Comment:
Added.

On Mon, Mar 28, 2016 at 12:00 PM, Brandon Vulaj <notifications@github.com>
wrote:

> In crud/src/main/java/com/redhat/lightblue/assoc/AnalyzeQuery.java
> <https://github.com/lightblue-platform/lightblue-core/pull/608#discussion_r57605869>
> :
>
> > +    private NearestCommonNode findNearestCommonNode(FieldTreeNode contextNode,FieldTreeNode fieldNode) {
> > +        int n=0;
> > +        FieldTreeNode trc=contextNode;
> > +        do {
> > +            if(isAnAncestor(trc,fieldNode)) {
> > +                return new NearestCommonNode(trc,n);
> > +            } else {
> > +                trc=trc.getParent();
> > +                n++;
> > +            }
> > +        } while(trc.getParent()!=null&&!(trc instanceof ArrayElement &&
> > +                                         ((ArrayElement)trc).getParent() instanceof ResolvedReferenceField ));
> > +        return new NearestCommonNode(null,n);
> > +    }
> > +
> > +    private QueryFieldInfo resolveField(Path clauseFieldName,
>
> Mind adding a comment here? There seems to be a method resolveField in
> almost every class in core.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/lightblue-platform/lightblue-core/pull/608/files/17418539ee2bfbf01e0fd0d96b8e9984280a6bbb#r57605869>
>